File: CSharp/Studies/Studies/RemoveDuplicates.cs
using System;

namespace Studies
{
    public static class RemoveDuplicatesExercises
    {
        /// <summary>
        /// https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/727/
        /// </summary>
        /// <returns>The length of the resized array</returns>
        /// <param name="arr">an array of sorted ints.</param>
        public static int RemoveDuplicatesFromSortedInts(ref int[] arr)
        {
            var shiftCount = 0;
            var shiftedArrLength = arr.Length;

            for(var i = 1; i < shiftedArrLength; i++)
            {
                if (arr[i] == arr[i-1])
                {
                    i--;
                    shiftCount++;
                    shiftedArrLength--;
                    Array.Copy(arr, i + 1, arr, i, arr.Length - (i + 1));
                }
            }

            Array.Resize(ref arr, shiftedArrLength);

            return arr.Length;
        }
    }   
}

File: js/lib/arrays/RemoveDuplicates.js
class RemoveDuplicates {
    
    /**
     * 
     * @param {Array} ints - a sorted set of numbers to remove duplicates from
     * @summary https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/727/
     * 
     * @returns new length of array after duplicates have been removed
     */
    static removeDuplicatesFromSortedInts(ints) {        
        for (let i = 1; i < ints.length; i ++) {
            if (ints[i] == ints[i-1]) {
                ints.splice(i, 1);
                i--;
            }
        }

        return ints.length;
    }

    /**
     * Recursive version
     * @param {Array} ints - a sorted set of numbers to remove duplicates from
     * @summary https://leetcode.com/explore/interview/card/top-interview-questions-easy/92/array/727/
     * 
     * @returns new length of array after duplicates have been removed
     */
    static removeDuplicatesFromSortedInts2(ints) {        
        return this.removeDuplicatesFromSortedIntsRecursion(ints, 0);
    }

    static removeDuplicatesFromSortedIntsRecursion(ints, idx) {
        if (idx >= ints.length) return ints.length;

        if (ints[idx] == ints[idx-1]) {
            ints.splice(idx, 1);
            idx--;
        }

        return this.removeDuplicatesFromSortedIntsRecursion(ints, ++idx);
    }
}

module.exports = RemoveDuplicates;